include/sampleflow/producers/delayed_rejection_mh.h

```diff
@@ -0,0 +1,164 @@
+         *     of the pair returned by the `perturb` argument is
+         *     $\frac{\pi_\text{proposal}(\tilde x|x)}
+         *           {\pi_\text{proposal}(x|\tilde x)}$.
+         *   </ol>
+         *   If samples are from some continuous space, say
+         *   ${\mathbb R}^n$, then the perturbation function is often
+         *   implemented by choosing $\tilde x$ from a neighborhood
+         *   of $x$. If, for example, $\tilde x$ is chosen with a probability
+         *   that only depends on the distance $\|\tilde x-\x\|$ as is often
+         *   done (e.g., uniformly in a disk of a certain radius around
+         *   $x$, or using a Gaussian probability distribution centered at
+         *   $x$), then the ratio returned as second argument is
+         *   $\frac{\pi_\text{proposal}(\tilde x|x)}
+         *           {\pi_\text{proposal}(x|\tilde x)}=1$.
+         * @param[in] n_samples The number of (new) samples to be produced
+         *   by this function. This is also the number of times the
+         *   signal is called that notifies Consumer objects that a new
+         *   sample is available.
+         * @param[in] max_delays The maximum number of delayed rejection stages.
+         */
+        void
+        sample (const OutputType &starting_point,
+                const std::function<double (const OutputType &)> &log_likelihood,
+                const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+                const types::sample_index n_samples,
+                const int max_delays);
+    };
+
+
+    template <typename OutputType>
+    void
+    DelayedRejectionMetropolisHastings<OutputType>::
+    sample (const OutputType &starting_point,
+            const std::function<double (const OutputType &)> &log_likelihood,
+            const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+            const types::sample_index n_samples,
+            const int max_delays)
+    {
+      // Make sure the flush_consumers() function is called at any point
+      // where we exit the current function.
+      Utilities::ScopeExit scope_exit ([this]()
+      {
+        this->flush_consumers();
+      });
+
+      std::mt19937 rng;
+      std::uniform_real_distribution<> uniform_distribution(0,1);
+
+      OutputType current_sample         = starting_point;
+      double     current_log_likelihood = log_likelihood (current_sample);
+
+      // Loop over the desired number of samples
+      for (types::sample_index i=0; i<n_samples; ++i)
+        {
+          // Initialize a vector to store rejected samples
+          std::vector<OutputType> rejected_samples;
+          // Initialize a bool to store whether a sample is accepted
+          bool accepted_sample = false;
+          // Delayed rejection loop
+          for (int delay_stage = 0; delay_stage < max_delays; delay_stage++)
```

Comment:
Conceptually, I would have expected that you always do at least one try, and that the number of "delays" indicate stepping from one try to another. So `max_delays==0` should result in one try, `max_delays==1` should result in two tries, etc. This would call for the following change:
```suggestion
          for (int delay_stage = 0; delay_stage <= max_delays; delay_stage++)
```

Reply:
We've also generally used the prefix-increment form:
```suggestion
          for (int delay_stage = 0; delay_stage < max_delays; ++delay_stage)
```